### PR TITLE
Uds support

### LINKED ISSUE #3092: Switch Client to use Unix Domain Sockets
Now that [we switched our HTTP client to undici](https://github.com/prisma/prisma/issues/2890) we can also start supporting UDS.

This should be able to be controlled with a flag to Prisma Client, so we can first test it in isolation and then later switch back to the old implementation to compare the performance or allow people to fall back to the proven implementation if there happens to be an incompatibility or problem. (The parameter does not have to be documented and officially be part of the API).

### PR DESCRIPTION
**Pull request by timsuchanek**
Closes #3092

### CODE REVIEW

File: src/packages/engine-core/src/NodeEngine.ts [hunk 1]
```diff
@@ -166,6 +172,12 @@ export class NodeEngine {
     )
     this.engineEndpoint = engineEndpoint
 
+    if (useUds && process.platform !== 'win32') {
+      this.socketPath = `/tmp/prisma-${getRandomString()}.sock`
```
Comment (Jolg42) [inline src/packages/engine-core/src/NodeEngine.ts:176]: What are the chances that the filesystem refuse access to `/tmp/` or if it's read only? I guess the socket would not initialize?

Should we check that first maybe like this?

```js
try {
	// R_OK | Flag indicating that the file can be read by the calling process.
	// W_OK | Flag indicating that the file can be written by the calling process.
    fs.accessSync(path.dirname(this.socketPath), fs.constants.R_OK | fs.constants.W_OK)
}
catch (err) {
    // err
}
```

File: src/packages/engine-core/src/NodeEngine.ts [hunk 2]
```diff
@@ -438,15 +450,18 @@ ${chalk.dim("In case we're mistaken, please report this to us 🙏.")}`)
...
       PRISMA_DML_PATH: this.datamodelPath,
-      PORT: String(this.port),
       RUST_BACKTRACE: '1',
       RUST_LOG: 'info',
     }
 
+    if (!this.useUds) {
+      env.PORT = String(this.port)
+    } else {
+      debug(`port: ${this.port}`)
```
Comment (Jolg42) [inline src/packages/engine-core/src/NodeEngine.ts:462]: Why is the debug only for useUds === true?

File: src/packages/engine-core/src/NodeEngine.ts [hunk 3]
```diff
@@ -565,7 +585,19 @@ ${chalk.dim("In case we're mistaken, please report this to us 🙏.")}`)
               json.target === 'query_engine::server' &&
               json.fields?.message.startsWith('Started http server')
             ) {
-              this.undici = new Undici(`http://localhost:${this.port}`)
+              if (this.useUds) {
+                this.undici = new Undici(
```
Comment (Jolg42) [inline src/packages/engine-core/src/NodeEngine.ts:589]: Should we change the default timeout here?
> socketTimeout: Number, the timeout after which a socket with active requests will time out. Monitors time between activity on a connected socket. Use 0 to disable it entirely. Default: 30e3 milliseconds (30s).

Comment (Jolg42) [reply src/packages/engine-core/src/NodeEngine.ts:589]: Just found out the 
```
requestTimeout: 0,
socketTimeout: 0,
```

File: src/packages/engine-core/src/undici.ts [hunk 4]
```diff
@@ -4,10 +4,13 @@ import { Pool } from 'undici'
 export class Undici {
   private pool: any
   private closed = false
-  constructor(url: string) {
+  constructor(url: any, moreArgs?: any) {
     this.pool = new Pool(url, {
```
Comment (Jolg42) [inline src/packages/engine-core/src/undici.ts:8]: Just found out about idleTimeout, not sure if we should change the default 🤔 

```
idleTimeout: Number, the timeout after which a socket without active requests will time out. Monitors time between activity on a connected socket. This value may be overriden by keep-alive hints from the server. Default: 4e3 milliseconds (4s).
```

**Review by Jolg42 — COMMENTED** [review]
Really nice!

**Comment by Jolg42** [comment]
Also it looks like test failed in CI

File: src/packages/engine-core/src/NodeEngine.ts (hunk 1 above)
Comment (timsuchanek) [reply src/packages/engine-core/src/NodeEngine.ts:176]: I'm not aware of any case, in which `/tmp` is not writable.
I would personally keep it simple for now.

File: src/packages/engine-core/src/NodeEngine.ts (hunk 2 above)
Comment (timsuchanek) [reply src/packages/engine-core/src/NodeEngine.ts:462]: Oops that's wrong, will fix

File: src/packages/engine-core/src/undici.ts (hunk 4 above)
Comment (timsuchanek) [reply src/packages/engine-core/src/undici.ts:8]: Ok we can keep the default there for now, will remove that :)

**Comment by timsuchanek** [comment]
> Also it looks like test failed in CI

Yep, that's why I'm setting up a Linux machine right now, because I can't reproduce the failing tests on Mac.

File: src/packages/engine-core/src/undici.ts (hunk 4 above)
Comment (Jolg42) [reply src/packages/engine-core/src/undici.ts:8]: Actually you didn't change it I think, but I was just thinking should we? 🤔

**Comment by timsuchanek** [comment]
UDS can be enabled with
```ts
new PrismaClient({ __internal: { useUds: true } })
```

**Comment by pantharshit00** [comment]
I tried it out and looks like it is working well :100:
image

image

**Pull request merged**